application/src/main/java/org/togetherjava/tjbot/commands/moderation/WarnCommand.java

```diff
@@ -0,0 +1,165 @@
+        if (!ModerationUtils.handleHasBotPermissions(ACTION_VERB, Permission.KICK_MEMBERS, bot,
+                guild, event)) {
+            return false;
+        }
+        if (!ModerationUtils.handleHasAuthorPermissions(ACTION_VERB, Permission.KICK_MEMBERS,
+                author, guild, event)) {
+            return false;
+        }
+        return ModerationUtils.handleReason(reason, event);
+    }
+
+    /**
+     * Handles {@code /warn user reason} command. Saves the value under the given user, given guild,
+     * given reason and adds one to the number of warns the user has.
+     * <p>
+     * This command can only be used by users with the {@code KICK_MEMBERS} permission.
+     *
+     * @param event the event of the command
+     */
+    @Override
+    public void onSlashCommand(@NotNull SlashCommandEvent event) {
+        OptionMapping userOption =
+                Objects.requireNonNull(event.getOption(USER_OPTION), "The option is null");
+        User target = userOption.getAsUser();
+        Member targetMember = userOption.getAsMember();
+        Member author = Objects.requireNonNull(event.getMember(), "The author is null");
+        Guild guild = Objects.requireNonNull(event.getGuild(), "The guild is null");
+        String reason = Objects.requireNonNull(event.getOption(REASON_OPTION), "The reason is null")
+            .getAsString();
+
+        Member bot = guild.getSelfMember();
+
+        if (!handleChecks(bot, author, targetMember, reason, guild, event)) {
+            return;
+        }
+
+        long userId = target.getIdLong();
+        dmUser(userId, reason, guild, event)
+            .map(hasSentDm -> sendFeedback(hasSentDm, target, author, reason))
+            .flatMap(event::replyEmbeds)
+            .queue();
+
+        long guildId = guild.getIdLong();
+        Optional<Integer> oldWarnAmount = database.read(context -> {
+            try (var select = context.selectFrom(WarnSystem.WARN_SYSTEM)) {
+                return Optional
+                    .ofNullable(select.where(WarnSystem.WARN_SYSTEM.USERID.eq(userId)
+                        .and(WarnSystem.WARN_SYSTEM.GUILD_ID.eq(guildId))).fetchOne())
+                    .map(WarnSystemRecord::getWarningAmount);
+            }
+        });
+
+        try {
+            database.write(context -> {
+                WarnSystemRecord warnSystemRecord = context.newRecord(WarnSystem.WARN_SYSTEM)
+                    .setUserid(target.getIdLong())
+                    .setGuildId(guildId)
+                    .setWarnReason(reason)
+                    .setIsWarned(true)
+                    .setWarningAmount(oldWarnAmount.orElse(0) + 1);
+                if (warnSystemRecord.update() == 0) {
+                    warnSystemRecord.insert();
+                }
+            });
+            logger.info("Saved the user '{}' to the warn system.", target.getAsTag());
+        } catch (Exception exception) {
+            exception.printStackTrace();
+        }
+    }
```

Comment:
The logic here is broken and will result in completely broken states in error cases.

The actual "warning" has to be made in the async chain after everything was successful. See the existing commands as a reference.

You can not just do it afterwards and in particular not in a blocking fashion.